br/pkg/restore/restorer.go

```diff
@@ -0,0 +1,348 @@
+							for _, f := range set.SSTFiles {
+								onProgress(int64(f.TotalKvs))
+							}
+						}
+					}()
+					err := s.fileImporter.Import(s.ectx, set)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					// TODO handle checkpoint
+					return nil
+				})
+		}
+	}
+	return nil
+}
+
+type MultiTablesRestorer struct {
+	eg               *errgroup.Group
+	ectx             context.Context
+	workerPool       *util.WorkerPool
+	fileImporter     ConcurrentlFileImporter
+	checkpointRunner *checkpoint.CheckpointRunner[checkpoint.RestoreKeyType, checkpoint.RestoreValueType]
+}
+
+func NewMultiTablesRestorer(
+	ctx context.Context,
+	fileImporter ConcurrentlFileImporter,
+	workerPool *util.WorkerPool,
+	checkpointRunner *checkpoint.CheckpointRunner[checkpoint.RestoreKeyType, checkpoint.RestoreValueType],
+) SstRestorer {
+	eg, ectx := errgroup.WithContext(ctx)
+	return &MultiTablesRestorer{
+		eg:               eg,
+		ectx:             ectx,
+		workerPool:       workerPool,
+		fileImporter:     fileImporter,
+		checkpointRunner: checkpointRunner,
+	}
+}
+
+func (m *MultiTablesRestorer) Close() error {
+	return m.fileImporter.Close()
+}
+
+func (m *MultiTablesRestorer) WaitUnitilFinish() error {
+	if err := m.eg.Wait(); err != nil {
+		summary.CollectFailureUnit("file", err)
+		log.Error("restore files failed", zap.Error(err))
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (m *MultiTablesRestorer) Restore(onProgress func(int64), batchFileSets ...BatchBackupFileSet) (err error) {
+	start := time.Now()
+	fileCount := 0
+	defer func() {
+		elapsed := time.Since(start)
+		if err == nil {
+			log.Info("Restore files", zap.Duration("take", elapsed))
+			summary.CollectSuccessUnit("files", fileCount, elapsed)
+		}
+	}()
+
+	log.Debug("start to restore files", zap.Int("files", fileCount))
+
+	if span := opentracing.SpanFromContext(m.ectx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("Client.RestoreSSTFiles", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+		m.ectx = opentracing.ContextWithSpan(m.ectx, span1)
+	}
+
+	for _, batchFileSet := range batchFileSets {
+		if m.ectx.Err() != nil {
+			log.Warn("Restoring encountered error and already stopped, give up remained files.",
+				logutil.ShortError(m.ectx.Err()))
+			// We will fetch the error from the errgroup then (If there were).
+			// Also note if the parent context has been canceled or something,
+			// breaking here directly is also a reasonable behavior.
+			break
+		}
+		filesReplica := batchFileSet
+		m.fileImporter.WaitUntilUnblock()
+		m.workerPool.ApplyOnErrorGroup(m.eg, func() (restoreErr error) {
+			fileStart := time.Now()
+			defer func() {
+				if restoreErr == nil {
+					log.Info("import files done", zap.Duration("take", time.Since(fileStart)))
+					onProgress(int64(len(filesReplica)))
+				}
+			}()
+			if importErr := m.fileImporter.Import(m.ectx, filesReplica...); importErr != nil {
```

Comment:
It seems we already managed the concurrenct execution. Why we still need `ConcurrentlFileImporter` here? Would you replace it with a normal one or add some comment here to explain?

Reply:
`ConcurrentFileImporter` only used in `MultiTablesRestorer`, which has a condition wait during restore. this make a better control of concurrency to make download balanced.

If we make this a default behavior for raw/txn ssts. Then it can be removed. but in this PR, I want to make the behavior keeps same as before.

Reply:
I'm wondering that given there is internal concurrency in the `ConcurrentFileImporter`, is the outer `workerPool` still needed?